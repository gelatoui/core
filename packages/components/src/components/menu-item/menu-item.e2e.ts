/* eslint-disable no-undef */
import { newE2EPage } from '@stencil/core/testing'

describe('glu-menu-item (E2E)', () => {
  it('renders correctly', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-menu-item type="label">Label Item</glu-menu-item>')

    const element = await page.find('glu-menu-item')

    expect(element).toHaveClass('hydrated')

    const label = element.shadowRoot.querySelector('glu-label')

    expect(label).toBeTruthy()
  })

  it('renders button type correctly', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-menu-item type="button">Click Me</glu-menu-item>')

    const element = await page.find('glu-menu-item')
    const button = element.shadowRoot.querySelector('glu-button')

    expect(button).toBeTruthy()
  })
})
