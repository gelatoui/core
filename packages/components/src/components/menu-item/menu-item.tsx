import { Component, Element, h, Host, Prop, State } from '@stencil/core'

/**
 * @component
 * @tag glu-menu-item
 * @shadow true
 *
 * @slot - The text of the menu item.
 * @slot right-text - The right text.
 * @slot dropdown-trigger - The trigger slot for dropdown type.
 * @slot dropdown-content - The dropdown content.
 */
@Component({
  tag: 'glu-menu-item',
  styleUrl: 'menu-item.css',
  shadow: true
})
export class GluMenuItem {
  /**
   * Defines the type of menu item.
   * @prop {string} type - The type of the menu item ('label', 'dropdown', 'button').
   */
  @Prop({ reflect: true }) readonly type: 'label' | 'dropdown' | 'button' = 'label'

  /**
   * URL to navigate to when clicking the menu item (used for 'label' and 'button' types).
   */
  @Prop() readonly href?: string

  /**
   * Specifies where to open the linked document. Only applicable if `href` is set.
   */
  @Prop() readonly target?: '_self' | '_blank' | '_parent' | '_top'

  /**
   * Specifies the relationship between the current document and the linked document. Only applicable if `href` is set.
   */
  @Prop() readonly rel?: string

  /**
   * A reference to the host element.
   */
  // eslint-disable-next-line no-undef
  @Element() menuItemElement!: HTMLGluMenuItemElement

  /**
   * State to track dropdown open/close.
   */
  @State() isOpen = false

  private toggleDropdown = () => {
    this.isOpen = !this.isOpen
  }

  render() {
    return (
      <Host class={`glu-menu-item glu-menu-item--${this.type}`} part="menu-item">
        {this.type === 'label' && (
          <glu-label class="label-container" onClick={this.toggleDropdown}>
            {this.href ?
              (
                <a href={this.href} target={this.target} rel={this.rel} class="label-container">
                  <slot></slot>
                </a>
              ) :
              (
                <slot></slot>
              )}
          </glu-label>
        )}

        {this.type === 'dropdown' && (
          <glu-dropdown isOpen={this.isOpen}>
            <div slot="dropdown-trigger" onClick={this.toggleDropdown} aria-expanded={this.isOpen}>
              <glu-label class="label-container">
                <slot></slot>
                <glu-icon name={this.isOpen ? 'chevron-up' : 'chevron-down'} size={16}></glu-icon>
              </glu-label>
            </div>
            <slot name="dropdown-content"></slot>
          </glu-dropdown>
        )}

        {this.type === 'button' && (
          <glu-button href={this.href} target={this.target} rel={this.rel}>
            <slot></slot>
          </glu-button>
        )}
      </Host>
    )
  }
}
