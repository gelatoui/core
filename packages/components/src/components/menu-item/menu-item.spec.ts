/* eslint-disable no-undef */
import { GluMenuItem } from './menu-item'

import { newSpecPage } from '@stencil/core/testing'

describe('glu-menu-item', () => {
  it('renders correctly as a label', async () => {
    const page = await newSpecPage({
      components: [GluMenuItem],
      html: '<glu-menu-item type="label">Label Item</glu-menu-item>'
    })

    expect(page.root).toMatchSnapshot()

    expect(page.root).toHaveClass('glu-menu-item')

    const label = page.root.ShadowRoot.querySelector('glu-label')

    expect(label).toBeTruthy()
  })

  it('renders correctly as a button', async () => {
    const page = await newSpecPage({
      components: [GluMenuItem],
      html: '<glu-menu-item type="button">Click Me</glu-menu-item>'
    })

    expect(page.root).toMatchSnapshot()

    expect(page.root).toHaveClass('glu-menu-item')

    const button = page.root.ShadowRoot.querySelector('glu-button')

    expect(button).toBeTruthy()
  })
})
