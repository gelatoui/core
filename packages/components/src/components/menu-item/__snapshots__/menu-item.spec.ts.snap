// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`glu-menu-item renders correctly as a button 1`] = `
<glu-menu-item class="glu-menu-item glu-menu-item--button" type="button">
  <template shadowrootmode="open">
    <glu-button>
      <slot></slot>
    </glu-button>
  </template>
  Click Me
</glu-menu-item>
`;

exports[`glu-menu-item renders correctly as a label 1`] = `
<glu-menu-item class="glu-menu-item glu-menu-item--label" type="label">
  <template shadowrootmode="open">
    <glu-label class="label-container">
      <slot></slot>
    </glu-label>
  </template>
  Label Item
</glu-menu-item>
`;
